lib/state-store.js

```diff
@@ -0,0 +1,311 @@
+      UpdateExpression: `REMOVE ${oldConsumers
+        .map((id, index) => `#a.#${index}`)
+        .join(', ')} SET #b = :x`,
+      ConditionExpression: `#b = :y`,
+      ExpressionAttributeNames: {
+        '#a': 'consumers',
+        '#b': 'version',
+        ...oldConsumers.reduce((obj, id, index) => ({ ...obj, [`#${index}`]: id }), {})
+      },
+      ExpressionAttributeValues: {
+        ':x': generate(),
+        ':y': version
+      }
+    };
+
+    try {
+      await client.update(params);
+      logger.debug(`Cleared old consumers: ${oldConsumers.join(', ')}`);
+    } catch (err) {
+      if (err.code !== 'ConditionalCheckFailedException') {
+        logger.error(err);
+        throw err;
+      }
+      logger.debug('Old consumers were cleared somewhere else.');
+    }
+  }
+
+  async registerConsumer() {
+    const privateProps = internal(this);
+    const { consumerId, consumerGroup, client, logger, streamName } = privateProps;
+
+    const heartbeat = new Date().toISOString();
+    const startedOn = new Date(Date.now() - uptime() * 1000).toISOString();
+
+    let params = {
+      Key: { consumerGroup, streamName },
+      UpdateExpression: 'SET #a.#b = :x',
+      ConditionExpression: 'attribute_not_exists(#a.#b)',
+      ExpressionAttributeNames: {
+        '#a': 'consumers',
+        '#b': consumerId
+      },
+      ExpressionAttributeValues: {
+        ':x': { appName, heartbeat, host, pid, startedOn }
+      }
+    };
+
+    try {
+      await client.update(params);
+      logger.debug(`The consumer "${consumerId}" is now registered.`);
+    } catch (err) {
+      if (err.code !== 'ConditionalCheckFailedException') {
+        logger.error(err);
+        throw err;
+      }
+    }
+
+    params = {
+      Key: { consumerGroup, streamName },
+      UpdateExpression: 'set #a.#b.#c = :x',
+      ExpressionAttributeNames: {
+        '#a': 'consumers',
+        '#b': consumerId,
+        '#c': 'heartbeat'
+      },
+      ExpressionAttributeValues: {
+        ':x': heartbeat
+      }
+    };
+
+    await client.update(params);
+  }
+
+  async getShardAndStreamState(shardId, shardData) {
+    const privateProps = internal(this);
+    const { logger, client, consumerGroup, streamName } = privateProps;
+
+    let streamState = await getStreamState(this);
+    let shardState = streamState.shards[shardId];
+    if (shardState !== undefined) return { shardState, streamState };
+
+    const params = {
+      Key: { consumerGroup, streamName },
+      UpdateExpression: 'SET #a.#b = :x',
+      ConditionExpression: 'attribute_not_exists(#a.#b)',
+      ExpressionAttributeNames: {
+        '#a': 'shards',
+        '#b': shardId
+      },
+      ExpressionAttributeValues: {
+        ':x': {
+          ...shardData,
+          checkpoint: null,
+          depleted: false,
+          leaseExpiration: null,
+          leaseOwner: null,
+          version: generate()
+        }
+      }
+    };
+
+    try {
+      await client.update(params);
```

Comment:
I'm confused why we're doing a client.update inside a "getShardAndStreamState" function.

Reply:
In case there's no state for the shard we use the shard definition data passed to initialize a state for it.